PCSX2/pcsx2
# Input-Recording: Add a new viewer/editor GUI Window

This is the last new GUI window on my list of ideas that input recording needs.  Two reasons for this:
1. It is apparently becoming more and more popular to create TASes (either in full, or partially) by editing the recording data itself, rather than actually "playing" the game.  This new window lays the basic groundwork for doing so, but also a location for future related tooling.
    - ie, you'll find a bunch of currently commented out menu options such as the ability to `export` and `import` frame data / bulk edit rows, etc.  These are all features I plan to follow up on.
2. It serves as a good tool to debug recording file problems / desync / etc.  Up until now I've been manually setting breakpoints or using the very spammy console log source.  In a vast majority of cases, a tool like this would have been _way_ nicer to work with.

![image](https://user-images.githubusercontent.com/13153231/111056725-56e40900-844f-11eb-8abd-5e039389a0e7.png)

### Testing Suggestions
- Make some recordings, ideally large ones to scale test the UI, load it up and try to:
  - Locate expected values
  - edit values / edit metadata (save and reload)
  - When saving make sure the original file isn't ruined (it should also be `.bak`'d if overwritten)
  - Change controller ports
  - Hide columns, these should persist across emulator restarts
- I've created a very basic caching strategy to make the grid not terribly slow.  Currently it will load 1000 frames before and after your current position.  However, if you quickly scroll through a multi-thousand frame recording, expect a bit of tearing (but nothing should break!)
